### PR TITLE
use double hmac comparision for tokens

### PR DESCRIPTION
**Pull request by suryagh**
Tokens are compared using normal strict-equal-to, this approach is vulnerable to [timing attacks](https://codahale.com/a-lesson-in-timing-attacks/). 

Node cannot perform constant-time-string comparison according to [relative recent discussion](https://github.com/nodejs/node-v0.x-archive/issues/8560#issuecomment-59793384) by node core team and the recommendation is to use [double HMAC comparison](https://www.nccgroup.trust/us/about-us/newsroom-and-events/blog/2011/february/double).

Fix: follows similar approach used in the [latest version of node](https://github.com/nodejs/node/pull/3073)

P.S. In case if you want to extract to an external module, there is a npm module [tsscmp](https://www.npmjs.com/package/tsscmp)

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+var crypto = require('crypto');
+
+function timeSafeCompare(a, b) {
+    a = String(a);
+    b = String(b);
+    var key = crypto.randomBytes(32);
+    var ah = crypto.createHmac('sha256', key).update(a).digest('base64');
+    var bh = crypto.createHmac('sha256', key).update(b).digest('base64');
+
+    return ((ah === bh) && (a === b));
```
Comment (indutny) [inline lib/utils.js]: Parens are not needed here.

File: test/utils.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+'use strict';
+
+var utils = require('../lib/utils'),
+    assert = require('assert');
+
+
+describe('utils', function () {
+    it('equal strings are true', function () {
+        assert.ok(utils.timeSafeCompare('127e6fbfe24a750e72930c220a8e138275656b8e5d8f48a98c3c92df2caba935',
+                                      '127e6fbfe24a750e72930c220a8e138275656b8e5d8f48a98c3c92df2caba935'),
```
Comment (indutny) [inline test/utils.js]: Alignment?

File: test/utils.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+                                      'equal strings not equal');
+    });
+
+    it('not equal strings are false', function () {
+        assert.ok(!utils.timeSafeCompare('alpha',
+                                        'beta'),
```
Comment (indutny) [inline test/utils.js]: Ditto.

File: lib/token.js [hunk 4]
```diff
@@ -27,7 +28,7 @@ function create(req, secretKey) {
                 return false;
 
             }
-            return token === tokenize(token.slice(0, LENGTH), req.session[secretKey]);
+            return utils.timeSafeCompare(token, tokenize(token.slice(0, LENGTH), req.session[secretKey]));
```
Comment (indutny) [inline lib/token.js]: idk, but do we care about 80 column limit?

**Comment by linkRace** [comment]
Looks solid to me, will be rolling this and others into a 1.5.0 soon.

**Comment by grawk** [comment]
Might we prefer using the `tsscmp` module to avoid the additional util and test file?

**Comment by linkRace** [comment]
@suryagh thoughts on `tsscmp`?

**Comment by suryagh** [comment]
@linkRace sounds good to me. I can update the PR.

**Comment by suryagh** [comment]
@linkRace @grawk updated to use `tsscmp`

**Comment by linkRace** [comment]
@grawk @indutny one more now that it's using tsscmp?

File: lib/csrf.js [hunk 5]
```diff
@@ -1,7 +1,8 @@
 'use strict';
 
 
-var token = require('./token');
+var token = require('./token'),
+    timeSafeCompare = require('tsscmp');
```
Comment (gabrielcsapo) [inline lib/csrf.js]: can we separate the declarations?

File: lib/token.js [hunk 6]
```diff
@@ -1,7 +1,8 @@
 'use strict';
 
 
-var crypto = require('crypto');
+var crypto = require('crypto'),
+    timeSafeCompare = require('tsscmp');
```
Comment (gabrielcsapo) [inline lib/token.js]: can we separate the declarations?

File: lib/csrf.js (hunk 5 above)
Comment (gabrielcsapo) [inline lib/csrf.js]: can we separate the declarations?

File: lib/token.js (hunk 6 above)
Comment (gabrielcsapo) [inline lib/token.js]: can we separate the declarations?

**Pull request merged**